From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Tue, 15 Apr 2025 10:24:04 +0200
Subject: [PATCH 008/225] ASoC: sprd: add support for single-channel DMA with
 interleaved data
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Advertise both INTERLEAVED and NONINTERLEAVED in the PCM driver and use
the selected access type to determine the effective number of DMA
channels. This makes the CPU DAI driver responsible for selecting the
right access type using snd_pcm_hw_constraint_mask64(), and fixes the
problem that the PCM driver currently advertises only the INTERLEAVED
access type but then splits the data between multiple DMA channels as
if it were NONINTERLEAVED.

When MCDT is used, there is only a single FIFO for two audio channels,
so the INTERLEAVED format is correct. However, it should then be sent to
the FIFO through one DMA channel. If each audio channel has a separate
destination, which may be the case if MCDT and the DSP are bypassed, the
NONINTERLEAVED format should be used instead.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 sound/soc/sprd/sprd-pcm-dma.c | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/sound/soc/sprd/sprd-pcm-dma.c b/sound/soc/sprd/sprd-pcm-dma.c
index 9bd4ec70d..963dcf64d 100644
--- a/sound/soc/sprd/sprd-pcm-dma.c
+++ b/sound/soc/sprd/sprd-pcm-dma.c
@@ -35,8 +35,9 @@ struct sprd_pcm_dma_private {
 
 static const struct snd_pcm_hardware sprd_pcm_hardware = {
 	.info = SNDRV_PCM_INFO_MMAP | SNDRV_PCM_INFO_MMAP_VALID |
-		SNDRV_PCM_INFO_INTERLEAVED | SNDRV_PCM_INFO_PAUSE |
-		SNDRV_PCM_INFO_RESUME | SNDRV_PCM_INFO_NO_PERIOD_WAKEUP,
+		SNDRV_PCM_INFO_INTERLEAVED | SNDRV_PCM_INFO_NONINTERLEAVED |
+		SNDRV_PCM_INFO_PAUSE | SNDRV_PCM_INFO_RESUME |
+		SNDRV_PCM_INFO_NO_PERIOD_WAKEUP,
 	.formats = SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE,
 	.period_bytes_min = 1,
 	.period_bytes_max = 64 * 1024,
@@ -205,6 +206,20 @@ static int sprd_pcm_hw_params(struct snd_soc_component *component,
 		return 0;
 	}
 
+	/*
+	 * In interleaved mode, all audio channels are passed through a single
+	 * DMA channel. Whether interleaved mode is used or not depends on the
+	 * constaints set by the CPU DAI.
+	 */
+	switch (params_access(params)) {
+	case SNDRV_PCM_ACCESS_MMAP_INTERLEAVED:
+	case SNDRV_PCM_ACCESS_RW_INTERLEAVED:
+		channels = 1;
+		break;
+	default:
+		break;
+	}
+
 	ret = sprd_pcm_request_dma_channel(component, substream, dma_params,
 					   channels);
 	if (ret)
